### PR TITLE
IGNITE-12857 Put custom objects via REST

### PR DESCRIPTION
**Pull request by xtern**

### CODE REVIEW

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java [hunk 1]
```diff
@@ -157,16 +160,15 @@
      *
      * @param hnd Handler.
      * @param authChecker Authentication checking closure.
-     * @param log Logger.
+     * @param ctx Kernal context.
      */
-    GridJettyRestHandler(GridRestProtocolHandler hnd, IgniteClosure<String, Boolean> authChecker, IgniteLogger log) {
+    GridJettyRestHandler(GridRestProtocolHandler hnd, IgniteClosure<String, Boolean> authChecker, GridKernalContext ctx) {
         assert hnd != null;
-        assert log != null;
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java:164]: Done

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java [hunk 2]
```diff
@@ -876,9 +805,11 @@ else if (cmd == CLUSTER_ACTIVE || cmd == CLUSTER_ACTIVATE)
             case EXECUTE_SQL_FIELDS_QUERY: {
                 RestQueryRequest restReq0 = new RestQueryRequest();
 
+                String cacheName = (String)params.get(CACHE_NAME_PARAM);
+
                 restReq0.sqlQuery((String)params.get("qry"));
 
-                restReq0.arguments(values(null, "arg", params).toArray());
+                restReq0.arguments(new Converter(cacheName).values( null, "arg", params).toArray());
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java]: Fixed

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+import org.apache.ignite.internal.binary.GridBinaryMarshaller;
+import org.apache.ignite.internal.processors.cache.GridCacheUtils;
+import org.apache.ignite.internal.processors.query.QueryTypeDescriptorImpl;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * JSON deserializer into the Ignite binary object.
+ */
+public class IgniteBinaryObjectJsonDeserializer extends JsonDeserializer {
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java]: Fixed

File: modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java [hunk 4]
```diff
@@ -2878,6 +3134,341 @@ public void ref(CircularRef ref) {
...
+                ", optional=" + optional +
+                ", list=" + list +
+                ", timestamp=" + timestamp +
+                ", bytes=" + Arrays.toString(longs) +
+                '}';
+        }
+    }
+
+    /** */
+    public enum COLOR {
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java:3288]: Fixed

File: modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java [hunk 5]
```diff
@@ -2878,6 +3134,341 @@ public void ref(CircularRef ref) {
...
+    /** Complex entity. */
+    @SuppressWarnings({"InnerClassMayBeStatic", "AssignmentOrReturnOfFieldWithMutableType"})
+    static class Complex implements Serializable {
+        /** */
+        @QuerySqlField(index = true)
+        private Integer id;
+
+        /** */
+        @QuerySqlField
+        private String string;
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java]: Dnne

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        this.ctx = ctx;
+    }
+
+    /** {@inheritDoc} */
+    @Override public BinaryObjectImpl deserialize(JsonParser parser, DeserializationContext dCtx) throws IOException {
+        String type = (String)dCtx.findInjectableValue(BINARY_TYPE_PROPERTY, null, null);
+        String cacheName = (String)dCtx.findInjectableValue(CACHE_NAME_PROPERTY, null, null);
+
+        assert type != null;
+        assert !ctx.marshallerContext().isSystemType(type);
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java]: Done
(Throwing IllegalArgumentException in this case)

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyObjectMapper.java [hunk 7]
```diff
@@ -264,4 +292,36 @@ private void writeException(Throwable e, JsonGenerator gen) throws IOException {
             }
         }
     };
+
+    /** Custom serializer for {@link java.sql.Timestamp}. */
+    private static final JsonSerializer<Timestamp> IGNITE_TIMESTAMP_SERIALIZER = new JsonSerializer<Timestamp>() {
+        /** {@inheritDoc} */
```
Comment (NSAmelchev) [inline modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyObjectMapper.java:298]: redundant javadoc

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java [hunk 8]
```diff
@@ -157,16 +160,16 @@
      *
      * @param hnd Handler.
      * @param authChecker Authentication checking closure.
-     * @param log Logger.
+     * @param ctx Kernal context.
      */
-    GridJettyRestHandler(GridRestProtocolHandler hnd, IgniteClosure<String, Boolean> authChecker, IgniteLogger log) {
+    GridJettyRestHandler(GridRestProtocolHandler hnd, IgniteClosure<String, Boolean> authChecker, GridKernalContext ctx) {
```
Comment (NSAmelchev) [inline modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java:165]: Too long line

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java [hunk 9]
```diff
@@ -1106,6 +1016,139 @@ private boolean credentials(Map<String, Object> params, String userParam, String
...
+            this(null);
+        }
+
+        /**
+         * Gets and converts values referenced by sequential keys, e.g. {@code key1...keyN}.
+         *
+         * @param type Optional value type.
+         * @param keyPrefix Key prefix, e.g. {@code key} for {@code key1...keyN}.
+         * @param params Parameters map.
+         * @return Values.
```
Comment (NSAmelchev) [inline modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java:1042]: Missed Throws javadoc

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java [hunk 10]
```diff
@@ -1106,6 +1016,139 @@ private boolean credentials(Map<String, Object> params, String userParam, String
         return null;
     }
 
+    /** */
```
Comment (NSAmelchev) [inline modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java:1019]: Add javadoc, please

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+            return deserialize0(cacheName, parentType.fieldTypeName(field), node, mapper);
+        }
+
+        /**
+         * @return Class descriptor for current binary type or {@code null} if the class was not found.
+         */
+        private @Nullable BinaryClassDescriptor binaryClassDescriptor() {
+            try {
+                return binType.context().descriptorForTypeId(false, binType.typeId(),null,false);
```
Comment (NSAmelchev) [inline modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java:320]: Add whitespace please

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyObjectMapper.java (hunk 7 above)
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyObjectMapper.java:298]: Thanks, 
I missed javadocs in similar places, actually.

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java (hunk 8 above)
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java]: Fixed, thanks.

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java [hunk 12]
```diff
@@ -1106,6 +1017,139 @@ private boolean credentials(Map<String, Object> params, String userParam, String
...
+        }
+
+        /**
+         * @param type Optional value type.
+         * @param s String to convert.
+         * @return Converted value.
+         * @throws IgniteCheckedException If failed to convert.
+         */
+        private Object convert(String type, String s) throws IgniteCheckedException {
+            if (F.isEmpty(type) || s == null)
```
Comment (NSAmelchev) [inline modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java:1070]: Can `s` be checked with `F.empty` too? Mark parametr as nullable, please.

File: modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java [hunk 13]
```diff
@@ -2778,6 +2991,87 @@ protected void initCache() {
...
+        return JSON_MAPPER.treeToValue(res, cls);
+    }
+
+    /**
+     * @param cacheName Cache name.
+     * @param params Query parameters.
+     * @return Json query result.
+     * @throws Exception If failed.
+     */
+    private JsonNode queryObject(String cacheName, String ... params) throws Exception {
```
Comment (NSAmelchev) [inline modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java:3052]: Unnecessary whitespace `String ...`

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java (hunk 9 above)
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java:1045]: Done

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java (hunk 10 above)
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java]: done

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java (hunk 11 above)
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java]: done

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java (hunk 12 above)
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/GridJettyRestHandler.java]: added Nullable

_Can s be checked with F.empty too?_
not sure,

File: modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java (hunk 13 above)
Comment (xtern) [reply modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java]: Done

File: modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java [hunk 14]
```diff
@@ -2878,6 +3163,276 @@ public void ref(CircularRef ref) {
...
+        /** */
+        GREEN,
+
+        /** */
+        BLUE
+    }
+
+    /** Complex entity. */
+    @SuppressWarnings({"InnerClassMayBeStatic", "AssignmentOrReturnOfFieldWithMutableType"})
+    static class Complex implements Serializable {
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java:3289]: Done.

File: modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+        @Nullable protected Object readValue(String field, JsonNode jsonNode) throws IOException {
+            JsonNodeType nodeType = jsonNode.getNodeType();
+
+            if (nodeType == JsonNodeType.BINARY)
+                return jsonNode.binaryValue();
+
+            if (nodeType == JsonNodeType.BOOLEAN)
+                return jsonNode.booleanValue();
+
+            Class<?> cls = qryFields.get(field.toUpperCase());
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/rest-http/src/main/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/IgniteBinaryObjectJsonDeserializer.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryClassDescriptor.java [hunk 16]
```diff
@@ -421,7 +421,7 @@ boolean isBinary() {
     /**
      * @return Described class.
      */
-    Class<?> describedClass() {
+    public Class<?> describedClass() {
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryClassDescriptor.java:424]: Can we use `BinaryUtils#resolveClass` instead of making this method public?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/query/GridQueryProcessor.java [hunk 17]
```diff
@@ -2949,6 +2949,17 @@ private String typeName(@Nullable String cacheName, String typeName) throws Igni
...
     }
 
+    /**
+     * Gets type descriptor for provided cache name and type name if type is still valid.
+     *
+     * @param cacheName Cache name.
+     * @param typeName Type name.
+     * @return Query type descriptor or {@code null} if descriptor was not found.
+     */
+    public @Nullable QueryTypeDescriptorImpl typeDescriptor(@Nullable String cacheName, String typeName) {
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/query/GridQueryProcessor.java:2959]: `QueryTypeDescriptorImpl` -> `GridQueryTypeDescriptor`?

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryClassDescriptor.java (hunk 16 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryClassDescriptor.java]: Fixed, thank you.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/query/GridQueryProcessor.java (hunk 17 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/query/GridQueryProcessor.java]: Done

File: modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java [hunk 18]
```diff
@@ -344,6 +355,252 @@ private void checkJson(String json, Person p) throws IOException {
...
+            "qry", "orgId = ?",
+            "arg1", String.valueOf(simple.getOrganizationId())
+        );
+
+        assertEquals(simple.getOrganizationId().intValue(), val.get("orgId").intValue());
+        assertEquals(simple.getSalary(), val.get("salary").doubleValue());
+        assertEquals(simple.getFirstName(), val.get("firstName").textValue());
+        assertEquals(simple.getLastName(), val.get("lastName").textValue());
+
+        initCache();
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java]: Done

File: modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java [hunk 19]
```diff
@@ -344,6 +355,252 @@ private void checkJson(String json, Person p) throws IOException {
...
+        assertTrue(Collection.class.isAssignableFrom(jcache().get(3).getClass()));
+
+        int[] ints = {1, 2, 3};
+        putObject(DEFAULT_CACHE_NAME, "4", ints, int[].class.getName());
+        assertTrue(Arrays.equals(ints, (int[])jcache().get(4)));
+
+        Character[] chars = {'a', 'b', 'c'};
+        putObject(DEFAULT_CACHE_NAME, "5", chars, Character[].class.getName());
+        assertTrue(Arrays.equals(chars, (Object[])jcache().get(5)));
+    }
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java:628]: Done

File: modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java [hunk 20]
```diff
@@ -344,6 +355,252 @@ private void checkJson(String json, Person p) throws IOException {
...
+        OuterClass newTypeUpdate = new OuterClass(-1, "update", 0.7d, list,
+            Timestamp.valueOf("2004-08-26 16:47:03.14"), new long[] {Long.MAX_VALUE, 0, Long.MAX_VALUE},
+            UUID.randomUUID(), IgniteUuid.randomUuid(), new OuterClass.OptionalObject("test"));
+
+        putObject(cache.getName(), "301", newTypeUpdate, valType);
+
+        assertEquals(newTypeUpdate, getObject(cache.getName(), "301", OuterClass.class));
+
+        // Check query result.
+        JsonNode res = queryObject(
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java]: Done

File: modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java [hunk 21]
```diff
@@ -344,6 +355,252 @@ private void checkJson(String json, Person p) throws IOException {
...
+
+        BinaryObject obj = parent.build();
+
+        // Registering "Parent" type.
+        jcache().put(2, obj);
+
+        // Adding another "Parent" object via REST.
+        JsonNode jsonNode = JSON_MAPPER.valueToTree(obj);
+
+        ((ObjectNode)jsonNode).put("id", Long.MIN_VALUE);
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/clients/src/test/java/org/apache/ignite/internal/processors/rest/JettyRestProcessorAbstractSelfTest.java]: Done
